src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -250,24 +235,126 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
             isAccessedBeforeAssignment = false;
         }
 
-        private BinaryExpressionSyntax GetLeftmostCondition(ExpressionSyntax condition)
+        private static bool TryGetTypeCheckParts(
+            SyntaxNode operand,
+            StatementSyntax statement,
+            out SyntaxNode variableDeclarator,
+            out SyntaxNode asExpression)
+        {
+            switch (operand.Kind())
+            {
+                case SyntaxKind.IdentifierName:
+                {
+                    // var x = e as T;
+                    // if (x != null) F(x);
+                    var identifier = (IdentifierNameSyntax)operand;
+                    var declarator = FindVariableDeclarator(identifier, statement);
+                    var initializerValue = declarator?.Initializer?.Value;
+                    if (initializerValue.IsKind(SyntaxKind.AsExpression) != true)
+                    {
+                        break;
+                    }
+
+                    variableDeclarator = declarator;
+                    asExpression = initializerValue;
+                    return true;
+                }
+
+                case SyntaxKind.SimpleAssignmentExpression:
+                {
+                    // T x;
+                    // if ((x = e as T) != null) F(x);
+                    var assignment = (AssignmentExpressionSyntax)operand;
+                    if (!assignment.Right.IsKind(SyntaxKind.AsExpression) ||
+                        !assignment.Left.IsKind(SyntaxKind.IdentifierName))
+                    {
+                        break;
+                    }
+
+                    var identifier = (IdentifierNameSyntax)assignment.Left;
+                    var declarator = FindVariableDeclarator(identifier, statement);
+                    if (declarator == null)
+                    {
+                        break;
+                    }
+
+                    variableDeclarator = declarator;
+                    asExpression = assignment.Right;
+                    return true;
+                }
+            }
+
+            variableDeclarator = null;
+            asExpression = null;
+            return false;
+        }
+
+        private static VariableDeclaratorSyntax FindVariableDeclarator(
+            IdentifierNameSyntax identifier, StatementSyntax usage)
```

Comment:
Should be TryFind since this can fail.  Maybe say "TryFind..InContainingBlock" as well.  That would help explain why we need to hand write this instead of just using the ILocalSymbol and getting the declaration from that.